File: app/src/main/java/com/ivanmagda/habito/view/model/HabitListItemViewModel.java
package com.ivanmagda.habito.view.model;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Color;

import com.ivanmagda.habito.R;
import com.ivanmagda.habito.models.Habit;
import com.ivanmagda.habito.models.ResetFrequency;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class HabitListItemViewModel {

    private Habit mHabit;
    private Context mContext;

    private static final SimpleDateFormat FORMAT = new SimpleDateFormat("MM/dd/yyyy", Locale.getDefault());

    public HabitListItemViewModel(Context context) {
        this.mContext = context;
    }

    public HabitListItemViewModel(Context context, Habit habit) {
        this.mContext = context;
        this.mHabit = habit;
    }

    public void setHabit(Habit habit) {
        this.mHabit = habit;
    }

    public int getBackgroundColor() {
        return mHabit.getRecord().getColor();
    }

    public String getHabitName() {
        return mHabit.getRecord().getName();
    }

    public int getHabitNameTextColor() {
        int color = mHabit.getRecord().getColor();
        if (color == Color.WHITE) {
            return mContext.getResources().getColor(R.color.primary_text);
        } else {
            return Color.WHITE;
        }
    }

    public String getResetFreq() {
        Resources resources = mContext.getResources();
        switch (mHabit.getRecord().getResetFreq()) {
            case ResetFrequency.DAY:
                return resources.getString(R.string.list_item_reset_today);
            case ResetFrequency.WEEK:
                return resetFreqStringWithParameter(ResetFrequency.WEEK);
            case ResetFrequency.MONTH:
                return resetFreqStringWithParameter(ResetFrequency.MONTH);
            case ResetFrequency.YEAR:
                return resetFreqStringWithParameter(ResetFrequency.YEAR);
            case ResetFrequency.NEVER:
                Date date = new Date(mHabit.getRecord().getCreatedAt());
                return resources.getString(R.string.list_item_reset_never, FORMAT.format(date));
            default:
                throw new IllegalArgumentException("Unsupported reset freq time");
        }
    }

    public String getScore() {
        final int score = mHabit.getRecord().getScore();
        final int target = mHabit.getRecord().getTarget();
        if (target > 0) {
            return String.valueOf(score) + "/"
                    + String.valueOf(target);
        } else {
            return String.valueOf(score);
        }
    }

    private String resetFreqStringWithParameter(String freqPeriod) {
        return mContext.getResources().getString(R.string.list_item_reset_freq, freqPeriod);
    }

}

File: app/src/main/java/com/ivanmagda/habito/view/model/HabitListItemViewModel.kt
package com.ivanmagda.habito.view.model

import android.content.Context
import android.graphics.Color
import com.ivanmagda.habito.R
import com.ivanmagda.habito.model.Habit
import com.ivanmagda.habito.model.ResetFrequency
import java.text.SimpleDateFormat
import java.util.*

class HabitListItemViewModel(val context: Context, var habit: Habit? = null) {

    val backgroundColor: Int
        get() = habit!!.record.color

    val habitName: String
        get() = habit!!.record.name

    // TODO: 'getColor(Int): Int' is deprecated.
    val habitNameTextColor: Int
        get() {
            val color = habit?.record?.color
            return if (color == Color.WHITE) {
                context.resources.getColor(R.color.primary_text)
            } else {
                Color.WHITE
            }
        }

    val resetFreq: String
        get() {
            val resources = context.resources
            return when (habit!!.record.resetFreq) {
                ResetFrequency.DAY -> resources.getString(R.string.list_item_reset_today)
                ResetFrequency.WEEK -> resetFreqStringWithParameter(ResetFrequency.WEEK)
                ResetFrequency.MONTH -> resetFreqStringWithParameter(ResetFrequency.MONTH)
                ResetFrequency.YEAR -> resetFreqStringWithParameter(ResetFrequency.YEAR)
                ResetFrequency.NEVER -> {
                    val date = Date(habit!!.record.createdAt)
                    resources.getString(R.string.list_item_reset_never, FORMAT.format(date))
                }
                else -> throw IllegalArgumentException("Unsupported reset freq time")
            }
        }

    val score: String
        get() {
            val score = habit!!.record.score
            val target = habit!!.record.target
            return if (target > 0) {
                (score.toString() + "/"
                        + target.toString())
            } else {
                score.toString()
            }
        }

    private fun resetFreqStringWithParameter(freqPeriod: String): String {
        return context.resources.getString(R.string.list_item_reset_freq, freqPeriod)
    }

    companion object {
        private val FORMAT = SimpleDateFormat("MM/dd/yyyy", Locale.getDefault())
    }
}
